### PR TITLE
Elaborate on advantages and disadvantages of Julia

### PR DESCRIPTION
**Pull request by LilithHafner**
I used data from https://julialang.org/assets/2023-julia-user-developer-survey.pdf#page=6 pages 6–8 to inform these additions.

@ViralBShah are the results of this survey available under the MIT liscence?

### CODE REVIEW

File: doc/src/index.md [hunk 1]
```diff
@@ -115,14 +115,25 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
...
     to [UTF-8](https://en.wikipedia.org/wiki/UTF-8)
   * Call C functions directly (no wrappers or special APIs needed)
   * Powerful shell-like capabilities for managing other processes
   * Lisp-like macros and other metaprogramming facilities
+
+### [Disadvantages of Julia](@id man-disadvantages-of-julia)
+
+  * It is difficult to generate self-contained binaries or libraries such as .exe, .dll, .so,
+    and impossible to generate small self-contained binaries with very fast runtimes
+  * Slow compile times - for example, it takes too long to generate the first plot
```
Comment (adienes) [inline doc/src/index.md:138]: while the sentiment is certainly well-known to be true, I think the wording with "too" here is a bit vague and subjective

File: doc/src/index.md [hunk 2]
```diff
@@ -115,14 +115,25 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
...
   * Efficient support for [Unicode](https://en.wikipedia.org/wiki/Unicode), including but not limited
     to [UTF-8](https://en.wikipedia.org/wiki/UTF-8)
   * Call C functions directly (no wrappers or special APIs needed)
   * Powerful shell-like capabilities for managing other processes
   * Lisp-like macros and other metaprogramming facilities
+
+### [Disadvantages of Julia](@id man-disadvantages-of-julia)
+
+  * It is difficult to generate self-contained binaries or libraries such as .exe, .dll, .so,
+    and impossible to generate small self-contained binaries with very fast runtimes
```
Comment (Keno) [inline doc/src/index.md:137]:
```suggestion
    and prohibitively difficult to generate small self-contained binaries with very fast runtimes
```

That said, I don't think this particular framing makes a ton of sense - it feels like we're saying we're giving up on these things, when that's very much not the case - they're being worked on and are not fundamental tradeoffs. I think a framing of "here's some limitations of the current implementation that you should be aware of when evaluating julia" is what we're going for here, but "disadvantages" doesn't really convey that for me.

**Comment by ViralBShah** [comment]
Sure, feel free to use it under the MIT license.

File: doc/src/index.md (hunk 2 above)
Comment (LilithHafner) [reply doc/src/index.md]: Does [Add header text](https://github.com/JuliaLang/julia/pull/52444/commits/3c901e63e4d23c44497caf3252ef38530427ae29) adequately address this? I do like the primary header being "disadvantages" as a counterpoint to "advantages" and I do think these things are actually disadvantages for folks considering using Julia (even if not for PL theorists)

**Comment by ViralBShah** [comment]
My main concern with this list is that it keeps changing as we release new Julia versions - this is a commitment in some way to keep the list updated. Also work in progress to address the issues should be somewhat reflected here (maybe issues, PRs, discourse discussions) - otherwise it gives the impression that nothing is being done.

**Comment by LilithHafner** [comment]
That's true; though I feel like it's not too much work to update this to the top few gripes from the community survey every year.

File: doc/src/index.md [hunk 3]
```diff
@@ -115,14 +115,29 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
...
   * Powerful shell-like capabilities for managing other processes
   * Lisp-like macros and other metaprogramming facilities
+
+### [Disadvantages of Julia](@id man-disadvantages-of-julia)
+
+As of 2023, the current implementation of Julia has some practical limitations that folks
+should be aware of. However, none of these are fundamental limitations of the language design
+and all are under active development:
+
+  * It is difficult to generate self-contained binaries or libraries such as .exe, .dll, .so,
```
Comment (KristofferC) [inline doc/src/index.md:140]: Why is it difficult? It is one command with PackageCompiler and for a binary all you need is a `@ccallable` in front of your main function. PackageCompiler will create a self contained bundle of all artifacts and libraries needed for you...

File: doc/src/index.md (hunk 1 above)
Comment (KristofferC) [reply doc/src/index.md:142]: > it takes too long to generate the first plot

With what package? With Plots.jl it is 2 seconds on my lap top. Is that too long? :shrug:

File: doc/src/index.md [hunk 4]
```diff
@@ -115,14 +115,29 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
...
+    and prohibitively difficult to generate small self-contained binaries with very fast runtimes
+  * Slow compile times - for example, it takes too long to generate the first plot
+  * When using deeply nested types or functions, error messages can be confusing and too verbose
```
Comment (KristofferC) [inline doc/src/index.md:143]: Should be improved with the new type parameter truncating functionality in stack traces?

**Comment by LilithHafner** [comment]
The reason I proposed this is that it seems a little sus to have an "Advantages of Julia" section but no mention of weaknesses, limitations, or why you would choose another language. There are obviously lots of solid reasons to choose a different language for all sorts of tasks and I think it makes sense to share some of them. At the very least if gives a bit more credibility to the "Advantages" section.

**Comment by ViralBShah** [comment]
The survey is a lagging indicator...

File: doc/src/index.md [hunk 5]
```diff
@@ -115,14 +115,29 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
 language. In addition to the above, some advantages of Julia over comparable systems include:
 
   * Free and open source ([MIT licensed](https://github.com/JuliaLang/julia/blob/master/LICENSE.md))
+  * Reliable built-in package manager that automatically installs compatible dependencies
+  * Talented, active, and welcoming community of developers
```
Comment (KristofferC) [inline doc/src/index.md:119]: I wouldn't put this in under

> some advantages of Julia over comparable systems include:

**Comment by KristofferC** [comment]
> The reason I proposed this is that it seems a little sus to have an "Advantages of Julia" section but no mention of weaknesses, limitations, or why you would choose another language.

I don't think it is "sus" to list features of a programming language in this way (a bit "markety") on an intro page. In fact, almost any language documentation I go into have something that is similar in spirit to this list (and none of them have a list of "disadvantages" that come from some somewhat old user survey.).

**Comment by simeonschaub** [comment]
Maybe there's a better way to frame this? More along the lines of "Julia is great for tons of application but for the following use cases you might run into limitations due to its current implementation. Be aware, but don't let those stop you – if you want to help tackle them reach out!"

**Comment by andrewjradcliffe** [comment]
By analogy with a résumé: the objective is to make a good first impression, hence, a succinct list of the positive qualities that even the average reader can understand.

Creating and maintaining a list of the negative qualities is likely to have the wrong effect. Experienced programmers might appreciate it (because they already have `m` gripes about the other `n` languages they know), but non-programmers might draw the wrong conclusions. If every language were as forthright as @LilithHafner is in this proposal, the world would be easier to navigate. Alas, it is not so.

File: doc/src/index.md [hunk 6]
```diff
@@ -115,14 +115,29 @@ Julia aims to create an unprecedented combination of ease-of-use, power, and eff
...
   * User-defined types are as fast and compact as built-ins
   * No need to vectorize code for performance; devectorized code is fast
   * Designed for parallelism and distributed computation
   * Lightweight "green" threading ([coroutines](https://en.wikipedia.org/wiki/Coroutine))
   * Unobtrusive yet powerful type system
+  * Reusability and composability—it is often possible to pass objects created
```
Comment (andrewjradcliffe) [inline doc/src/index.md:125]: To appreciate the full implication, this statement requires understanding of the Julia language (and, likely, a background in programming). Suggestion: re-phrase or eliminate

File: doc/src/index.md (hunk 5 above)
Comment (andrewjradcliffe) [inline doc/src/index.md:119]: Agree with @KristofferC. I would add that it is also difficult to quantify, so probably best let the links provided at the top of this file speak for the quality of the community.

File: doc/src/index.md (hunk 1 above)
Comment (LilithHafner) [reply doc/src/index.md:142]: If you include precompile time, Plots.jl takes 11 minutes to give a first plot on a pinebook pro. That's longer than it takes to install python and matplotlib, look up a hello world example code in that plotting system, write code to plot something, and plot it.

**Comment by LilithHafner** [comment]
This seems like an unpopular idea.

**Pull request closed without merging**